vm/src/vm.rs

```diff
@@ -1013,6 +1020,42 @@ impl VirtualMachine {
             })
     }
 
+    pub fn decode(
```

Comment:
There is a lot of code duplication between `decode` and `encode` how about creating some inner function?